Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/triangulate_faces.h

```diff
@@ -533,6 +579,159 @@ bool triangulate_faces(PolygonMesh& pmesh)
+
+  typedef typename GetVertexPointMap<PolygonMesh, NamedParameters>::const_type VPM;
+  const VPM vpm = parameters::choose_parameter(
+    parameters::get_parameter(np, internal_np::vertex_point),
+    get_const_property_map(boost::vertex_point, pmesh));
+
+  typedef Halfedge_around_face_circulator<PolygonMesh> Hedge_around_face_circulator;
+  Hedge_around_face_circulator circ(border_halfedge, pmesh);
+  Hedge_around_face_circulator done(circ);
+
+  // Getting the hole boundary.
+  std::vector<Point_3> polyline_3d;
+  do {
+    polyline_3d.push_back(get(vpm, target(*circ, pmesh)));
+  } while (++circ != done);
+  CGAL_assertion(
+    polyline_3d.size() >= 3);
+
+  // Compute an average normal of the hole face.
+  const Squared_length_3 squared_length_3 =
+    traits.compute_squared_length_3_object();
+  const Collinear_3 collinear_3 =
+    traits.collinear_3_object();
+
+  FT x = FT(0), y = FT(0), z = FT(0);
+  std::size_t num_normals = 0;
+  const Point_3& ref_point = polyline_3d[0];
+  for (std::size_t i = 1; i < polyline_3d.size() - 1; ++i) {
+    const std::size_t ip = i + 1;
+
+    const Point_3& p1 = ref_point; // 3 points, which form a triangle
+    const Point_3& p2 = polyline_3d[i];
+    const Point_3& p3 = polyline_3d[ip];
+
+    // Skip in case we have collinear points.
+    if (collinear_3(p1, p2, p3))
+      continue;
+
+    // Computing the normal of a triangle.
+    Vector_3 tri_normal = CGAL::normal(p1, p2, p3);
+    const FT tri_normal_length = static_cast<FT>(
+      CGAL::sqrt(CGAL::to_double(squared_length_3(tri_normal))));
+    CGAL_assertion(tri_normal_length > FT(0));
+    tri_normal /= tri_normal_length;
+
+    x += tri_normal.x(); y += tri_normal.y(); z += tri_normal.z();
+    ++num_normals;
+  }
+
+  CGAL_assertion(num_normals >= 1);
+  x /= static_cast<FT>(num_normals);
+  y /= static_cast<FT>(num_normals);
+  z /= static_cast<FT>(num_normals);
+
+  // Setting the final normal.
+  const Vector_3 normal = Vector_3(x, y, z);
+  const FT normal_length = static_cast<FT>(
+    CGAL::sqrt(CGAL::to_double(squared_length_3(normal))));
+  const Vector_3 avg_normal = normal / normal_length;
+
+  // Checking the hole simplicity.
+  typedef Triangulation_2_projection_traits_3<Traits> P_traits;
+  const P_traits p_traits(avg_normal);
+  const bool is_simple =
+    is_simple_2(polyline_3d.begin(), polyline_3d.end(), p_traits);
+  // if (!is_simple) // temporarily removed
+  //   return triangulate_hole(
+  //     pmesh, border_halfedge, out, np);
+
+  // Adding the new face.
+  typedef typename boost::graph_traits<PolygonMesh>::face_descriptor face_descriptor;
+  face_descriptor new_face = add_face(pmesh);
+  set_halfedge(new_face, border_halfedge, pmesh);
+  do {
+    set_face(*circ, new_face, pmesh);
+  } while (++circ != done);
+
+  // Triangulating.
+  const bool use_cdt = true;
+  internal::Triangulate_modifier<PolygonMesh, VPM, Traits> modifier(vpm, traits);
+  const bool success_with_cdt_2 =
```

Comment:
OK here is the part I missed in my previous review. The intersection of constraints as well has the vertices being identical upon projects is handled in the function `triangulate_face()`.

Reply:
Yes.